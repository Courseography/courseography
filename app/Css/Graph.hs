{-|
    Module      : Css.Graph
    Description : Defines the CSS for the graph page.
-}
module Css.Graph
    (graphStyles) where

import Clay
import Prelude hiding ((**))
import Data.Monoid
import Css.Constants

-- |Defines all CSS for the graph page.
graphStyles :: Css
graphStyles = do
    graphContainer
    sidebarCSS
    nodeCSS
    pathCSS
    resetCSS
    titleCSS
    regionCSS

{- nodeCSS
 - Generates CSS for nodes in the graph. -}
nodeCSS :: Css
nodeCSS = "g" ? do
    "text" ? do
        userSelect none
        "-webkit-touch-callout" -: "none"
        "-webkit-user-select" -: "none"
        "-khtml-user-select" -: "none"
        "-moz-user-select" -: "none"
        "-ms-user-select" -: "none"

    ".node" & do
        cursor pointer
        "text" ? do
            fontSize (pt nodeFontSize)
            faded
            stroke "none"
            "text-anchor" -: "middle"
        -- For nodes in draw tab
        "data-active" @= "active" & do
            "rect" <? do
                wideStroke
            "text" <? do
                fullyVisible
        "data-active" @= "overridden" & do
            "rect" <? do
                wideStroke
                strokeRed
            "text" <? do
                fullyVisible
        "data-active" @= "inactive" & do
            "rect" <? do
                faded
                strokeDashed
        "data-active" @= "takeable" & do
            "rect" <? do
                semiVisible
            "text" <? do
                semiVisible
        "data-active" @= "missing" & do
            "rect" <> "ellipse" <? do
                wideStroke
                strokeRed
            "text" <? do
                fullyVisible
        "data-active" @= "unselected" & do
            "rect" <? do
                wideStroke
                faded
        "rect" <? do
            stroke "black"
    ".hybrid" & do
        cursor cursorDefault
        "text" <? do
            stroke "none"
            fill "white"
            fontSize (pt hybridFontSize)
            "text-anchor" -: "middle"
        "rect" <? do
            fill "#888888"
            stroke "black"
    ".active" & do
        "rect" <? do
            wideStroke
        "text" <? do
            fullyVisible
    ".overridden" & do
        "rect" <? do
            wideStroke
            strokeRed
        "text" <? do
            fullyVisible
    ".inactive" & do
        "rect" <? do
            faded
            strokeDashed
    ".takeable" & do
        "rect" <? do
            semiVisible
        "text" <? do
            semiVisible
    ".missing" & do
        "rect" <> "ellipse" <? do
            wideStroke
            strokeRed
        "text" <? do
            fullyVisible
    ".bool" & do
        cursor cursorDefault
        "data-active" @= "active" & do
            "ellipse" <? do
                fill "none"
                stroke "black"
        "data-active" @= "overridden" & do
            "ellipse" <? do
                fill "white"
                strokeRed
        "data-active" @= "inactive" & do
            "ellipse" <? do
                fill lightGrey
                faded
                strokeDashed
                stroke "black"
        "data-active" @= "takeable" & do
            "ellipse" <? do
                fill lightGrey
                stroke "black"
        "data-active" @= "missing" & do
            "ellipse" <? do
                fill "white"
                strokeRed
        -- For the React graph
        ".active" & do
            "ellipse" <? do
                fill "none"
                stroke "black"
        ".overridden" & do
            "ellipse" <? do
                fill "white"
                strokeRed
        ".inactive" & do
            "ellipse" <? do
                fill lightGrey
                faded
                strokeDashed
                stroke "black"
        ".takeable" & do
            "ellipse" <? do
                fill lightGrey
                stroke "black"
        ".missing" & do
            "ellipse" <? do
                fill "white"
                strokeRed
        "text" <? do
            fontFamily ["Trebuchet MS", "Arial"] [sansSerif]
            fontWeight bold
            stroke "none"
            fontSize (pt boolFontSize)
            "text-anchor" -: "middle"

{- pathCSS
 - Generates CSS for paths between nodes
 - in the graph. -}
pathCSS :: Css
pathCSS = "path" ? do
    fill "none"
    "data-active" @= "drawn" & do
        faded
        wideStroke
    -- For the React graph
    ".takeable" & do
        strokeDashed
        stroke "black"
    ".inactive" & do
        faded
        strokeDashed
        stroke "black"
    ".active" & do
        opacity 1
        "stroke-width" -: "2px"
        stroke "black"
    ".missing" & do
        strokeRed
        strokeDashed

{- resetCSS
 - Generates CSS for the reset feature
 - in the graph. -}
resetCSS :: Css
resetCSS = "#resetButton" ? do
    fill "#990000"
    cursor pointer
    wideStroke
    "stroke" -: "#404040"
    "text" <? do
        fill "white"

{- graphContainer
 - Generates CSS for the main division of
 - the page containing the graph. -}
graphContainer :: Css
graphContainer = do
    "#react-graph" ? do
        "width" -: "calc(100% - 40px)"
        height100
        overflow hidden
        margin0
        display inlineBlock
        position absolute
        textAlign $ alignSide sideCenter
        left (px 40)
    "#react-graphRootSVG" ? do
        width100
        height100
        stroke "black"
        "stroke-linecap" -: "square"
        "stroke-miterlimit" -: "10"
        "shape-rendering" -: "geometricPrecision"
    ".highlight-nodes" ? do
        backgroundColor grey
    ".graph-control-button" ? do
        textAlign $ alignSide sideCenter
        backgroundColor white
        border solid (px 1) purple10
        "border-radius" -: "6px"
        "outline" -: "none"
        fontSize (px 14)
        fontWeight bold
        fontColor purple10
        cursor pointer
        display inlineBlock
        width (px 25)
        height (px 25)
        position absolute
        ":hover" & do
            backgroundColor purple6
            fontColor white
        ":active" & do
            backgroundColor purple10
            fontColor white
        ":disabled" & do
            border solid (px 1) grey6
            backgroundColor grey2
            fontColor grey6
    "#zoom-in-button" ? do
        top (px 85)
        right (px 30)
    "#zoom-out-button" ? do
        top (px 85)
        right (px 57)
    "#pan-up-button" ? do
        top (px 4)
        right (px 43)
    "#pan-down-button" ? do
        top (px 57)
        right (px 43)
    "#pan-right-button" ? do
        top (px 30)
        right (px 30)
    "#pan-left-button" ? do
        top (px 30)
        right (px 57)
    "#reset-button" ? do
        width (px 52)
        top (px 111)
        right (px 30)

sidebarCSS :: Css
sidebarCSS = do
    "#fce" ? do
        height (px 40)
        width (pct 100)
        border solid (px 1) black
    "#fcecount" ? do
        width (pct 53)
        height (px 40)
        float floatLeft
        backgroundColor purple7
        display none
        textAlign $ alignSide sideCenter
        paddingLeft (px 15)
        paddingTop (px 5)
        fontSize (px 18)
    "#reset" ? do
        textAlign $ alignSide sideCenter
        float floatLeft
        width (pct 47)
        height (px 40)
        display none
        border solid (px 1) black
        cursor pointer
        ":hover" & do
            backgroundColor grey1
            fontColor white
    "#container" ? do
        "height" -: "calc(100% - 100px)"
        width100
        position relative
    "#sidebar" ? do
        display inlineBlock
        width (px 40)
        height100
        float floatLeft
        backgroundColor purple8
        position absolute
        paddingLeft (px 23)
    "#sidebar-button" ? do
        cursor pointer
        display inlineBlock
        width (px 40)
        height100
        float floatLeft
        backgroundColor purple10
        position absolute
        border solid (px 1) black
        ":hover" & do
            backgroundColor purple6
    "#sidebar-icon" ? do
        width (px 30)
        height (px 50)
        paddingTop (px 20)
        paddingLeft (px 2)
        position absolute
        top (pct 40)
        left (px 4)
    "#focuses-nav, #graph-nav" ? do
        cursor pointer
    "#sidebar-nav" ? do
        width100
        fontSize (px 13)
        backgroundColor purple6
        border solid (px 1) grey2
        "box-shadow" -: "0 2px 2px -1px rgba(0, 0, 0, 0.055)"
        display block
        overflow hidden
        ul ? do
            width100
            margin0
            li ? do
                "list-style-type" -: "none"
                display inlineBlock
                width (pct 48)
                "-webkit-transition" -: "all 0.2s"
                "-moz-transition" -: "all 0.2s"
                "-ms-transition" -: "all 0.2s"
                "-o-transition" -: "all 0.2s"
                "transition" -: "all 0.2s"
                ":hover" & do
                    "background-color" -: "#5C497E !important"
                    a ? do
                        "color" -: "white !important"
                a ? do
                    color black
                    display inlineBlock
                    lineHeight (px 30)
                    alignCenter
                    width (pct 95)
                    textDecoration none
    "#focuses, #graphs" ? do
        marginTop (px 25)
        marginLeft (px 25)
        height (pct 86)
        display none
        overflowY scroll
    ".focus" ? do
        display block
        cursor pointer
        fontSize (px 20)
        border solid (px 1) black
        alignCenter
        width (pct 90)
        backgroundColor white
        "border-radius" -: "6px"
        ":hover" & do
            backgroundColor grey2
    "#close-focus" ? do
        display block
        cursor pointer
        backgroundColor purple6
        fontSize (px 20)
        border solid (px 1) black
        textAlign $ alignSide sideCenter
        width (pct 90)
    ".spotlight" & do
        fill "white"
        stroke "none"
    ".details" & do
        border solid (px 1) black
        width (pct 90)
        height (px 0)
        marginBottom (px 7)
        overflow auto
        fontSize (px 14)
        fontColor white
        paddingLeft (px 5)
        paddingRight (px 5)
    ".active" & do
        backgroundColor purple8
    ".disabled" & do
        backgroundColor grey2
        pointerEvents none
    ".graph-button" & do
        display block
        cursor pointer
        fontSize (px 20)
        border solid (px 1) black
        "border-radius" -: "6px"
        textAlign $ alignSide sideCenter
        width (pct 90)
        backgroundColor white
        marginBottom (px 20)
        ":hover" & do
            backgroundColor grey2
    ".flip" & do
        transform $ scaleX (-1)


{- titleCSS
 - Generates CSS for the title. -}
titleCSS :: Css
titleCSS = "#svgTitle" ? do
    fontSize $ em 2.5
    fontWeight bold
    fontFamily ["Bitter"] [serif]
    fontStyle italic
    fill titleColour

{- regionCSS
 - Generates CSS for focus regions in the graph. -}
regionCSS :: Css
regionCSS = do
    ".region-label" ? do
        fontSize (pt regionFontSize)
    ".region" ? do
        "fill-opacity" -: "0.25"
