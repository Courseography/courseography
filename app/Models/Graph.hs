module Models.Graph
    (getGraph,
    saveGraphJSON) where

import Config (runDb)
import Control.Monad.IO.Class (liftIO)
import Data.Aeson (Value, decode, object, toJSON)
import Data.List (partition)
import qualified Data.Text as T (Text)
import Database.DataType (ShapeType (BoolNode, Hybrid, Node))
import Database.Persist.Sqlite (Entity, PersistEntity, PersistValue (PersistInt64), SqlPersistM,
                                entityKey, entityVal, insert, insertMany_, keyToValues, selectFirst,
                                selectList, (<-.), (==.))
import Database.Tables hiding (paths, shapes, texts)
import Happstack.Server (Response, ServerPart, lookBS, lookText', toResponse)
import Svg.Builder (buildEllipses, buildPath, buildRect, intersectsWithShape)

getGraph :: T.Text -> IO (Maybe Value)
getGraph graphName = runDb $ do
    graphEnt :: (Maybe (Entity Graph)) <- selectFirst [GraphTitle ==. graphName] []
    case graphEnt of
        Nothing -> return Nothing
        Just graph -> do
            let gId = entityKey graph
            sqlTexts    :: [Entity Text] <- selectList [TextGraph ==. gId] []
            sqlRects    :: [Entity Shape] <- selectList
                                                 [ShapeType_ <-. [Node, Hybrid],
                                                  ShapeGraph ==. gId] []
            sqlEllipses :: [Entity Shape] <- selectList
                                                 [ShapeType_ ==. BoolNode,
                                                  ShapeGraph ==. gId] []
            sqlPaths    :: [Entity Path] <- selectList [PathGraph ==. gId] []

            let
                keyAsInt :: PersistEntity a => Entity a -> Integer
                keyAsInt = fromIntegral . (\(PersistInt64 x) -> x) . head . keyToValues . entityKey

                graphtexts          = map entityVal sqlTexts
                rects          = zipWith (buildRect graphtexts)
                                         (map entityVal sqlRects)
                                         (map keyAsInt sqlRects)
                ellipses       = zipWith (buildEllipses graphtexts)
                                         (map entityVal sqlEllipses)
                                         (map keyAsInt sqlEllipses)
                graphpaths     = zipWith (buildPath rects ellipses)
                                         (map entityVal sqlPaths)
                                         (map keyAsInt sqlPaths)
                (regions, _)   = partition pathIsRegion graphpaths
                regionTexts    = filter (not .
                                         intersectsWithShape (rects ++ ellipses))
                                        graphtexts

                response = object [
                        ("texts", toJSON $ graphtexts ++ regionTexts),
                        ("shapes", toJSON $ rects ++ ellipses),
                        ("paths", toJSON $ graphpaths ++ regions),
                        ("width", toJSON $ graphWidth $ entityVal graph),
                        ("height", toJSON $ graphHeight $ entityVal graph)
                    ]

            return (Just response)

-- | Inserts SVG graph data into Texts, Shapes, and Paths tables
saveGraphJSON :: ServerPart Response
saveGraphJSON = do
    jsonStr <- lookBS "jsonData"
    nameStr <- lookText' "nameData"
    let jsonObj = decode jsonStr :: Maybe SvgJSON
    case jsonObj of
        Nothing -> return $ toResponse ("Error" :: String)
        Just (SvgJSON texts shapes paths) -> do
            _ <- liftIO $ runDb $ insertGraph nameStr texts shapes paths
            return $ toResponse ("Success" :: String)
    where
        insertGraph :: T.Text -> [Text] -> [Shape] -> [Path] -> SqlPersistM ()
        insertGraph nameStr_ texts shapes paths = do
            gId <- insert $ Graph nameStr_ 256 256 False
            insertMany_ $ map (\text -> text {textGraph = gId}) texts
            insertMany_ $ map (\shape -> shape {shapeGraph = gId}) shapes
            insertMany_ $ map (\path -> path {pathGraph = gId}) paths
