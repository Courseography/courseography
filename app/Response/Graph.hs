module Response.Graph
    (graphResponse) where

import           Text.Blaze ((!))
import qualified Text.Blaze.Html5 as H
import qualified Text.Blaze.Html5.Attributes as A
import Happstack.Server
import MasterTemplate
import Scripts

graphResponse :: ServerPart Response
graphResponse =
   ok $ toResponse $
    masterTemplate "Courseography - Graph"
                []
                (do
                    header "graph"
                    H.div ! A.id "container" $ do
                        -- TODO: remove this inner div
                        H.div ! A.id "react-graph" ! A.class_ "react-graph" $ ""
                    disclaimer
                )
                graphScripts
