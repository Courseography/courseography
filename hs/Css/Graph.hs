{-# LANGUAGE OverloadedStrings #-}

module Css.Graph
    (graphStyles) where

import Clay
import Prelude hiding ((**))
import Data.Monoid
import Css.Constants

{- graphStyles
 - Generates all CSS for the graph page. -}
graphStyles :: Css
graphStyles = do
    graphContainer
    sidebarCSS
    nodeCSS
    pathCSS
    resetCSS
    titleCSS
    regionCSS

{- nodeCSS
 - Generates CSS for nodes in the graph. -}
nodeCSS :: Css
nodeCSS = "g" ? do
    "text" ? do
        userSelect none
        "-webkit-touch-callout" -: "none"
        "-webkit-user-select" -: "none"
        "-khtml-user-select" -: "none"
        "-moz-user-select" -: "none"
        "-ms-user-select" -: "none"
    ".node" & do
        cursor pointer
        "text" ? do
            fontSize (pt nodeFontSize)
            faded
            stroke "none"
            "text-anchor" -: "middle"
        "data-active" @= "active" & do
            "rect" <? do
                wideStroke
            "text" <? do
                fullyVisible
        "data-active" @= "overridden" & do
            "rect" <? do
                wideStroke
                strokeRed
            "text" <? do
                fullyVisible
        "data-active" @= "inactive" & do
            "rect" <? do
                faded
                strokeDashed
        "data-active" @= "takeable" & do
            "rect" <? do
                semiVisible
            "text" <? do
                semiVisible
        "data-active" @= "missing" & do
            "rect" <> "ellipse" <? do
                wideStroke
                strokeRed
            "text" <? do
                fullyVisible
        "data-active" @= "unlit" & do
            "rect" <?
                faded
        "data-active" @= "unselected" & do
            "rect" <? do
                wideStroke
                faded
        -- For nodes in draw tab
        "data-group" @= "red" & do
            "rect" <? do
                fill dRed
        "data-group" @= "blue" & do
            "rect" <? do
                fill dBlue
        "data-group" @= "green" & do
            "rect" <? do
                fill dGreen
        "data-group" @= "purple" & do
            "rect" <? do
                fill dPurple
        "rect" <? do
            stroke "black"
    ".hybrid" & do
        cursor cursorDefault
        "text" <? do
            stroke "none"
            fill "white"
            fontSize (pt hybridFontSize)
            "text-anchor" -: "middle"
        "rect" <? do
            fill "#888888"
            stroke "black"
    ".bool" & do
        cursor cursorDefault
        "data-active" @= "active" & do
            "ellipse" <? do
                fill "none"
                stroke "black"
        "data-active" @= "overridden" & do
            "ellipse" <? do
                fill "white"
                strokeRed
        "data-active" @= "inactive" & do
            "ellipse" <? do
                fill lightGrey
                faded
                strokeDashed
                stroke "black"
        "data-active" @= "takeable" & do
            "ellipse" <? do
                fill lightGrey
                stroke "black"
        "data-active" @= "missing" & do
            "ellipse" <? do
                fill "white"
                strokeRed
        "data-active" @= "unlit" & do
            wideStroke
            strokeRed
        "text" <? do
            fontFamily ["Trebuchet MS", "Arial"] [sansSerif]
            fontWeight bold
            stroke "none"
            fontSize (pt boolFontSize)
            "text-anchor" -: "middle"

{- pathCSS
 - Generates CSS for paths between nodes
 - in the graph. -}
pathCSS :: Css
pathCSS = "path" ? do
    fill "none"
    "data-active" @= "takeable" & do
        strokeDashed
    "data-active" @= "inactive" & do
        faded
        strokeDashed
    "data-active" @= "active" & do
        opacity 1
        "stroke-width" -: "2px"
    "data-active" @= "missing" & do
        strokeRed
        strokeDashed
    "data-active" @= "drawn" & do
        faded
        wideStroke


{- resetCSS
 - Generates CSS for the reset feature
 - in the graph. -}
resetCSS :: Css
resetCSS = "#resetButton" ? do
    fill "#990000"
    cursor pointer
    wideStroke
    "stroke" -: "#404040"
    "text" <? do
        fill "white"

{- graphContainer
 - Generates CSS for the main division of
 - the page containing the graph. -}
graphContainer :: Css
graphContainer = do
    "#graph" ? do
        width (px 1195)
        minHeight (px 700)
        height (px 700)
        overflow hidden
        margin (px 10) (px 10) (px 10) (px 10)
        display inlineBlock
        position absolute
        textAlign $ alignSide sideCenter
        "left" -: "40px"
    "#graphRootSVG" ? do
        width100
        height100
        stroke "black"
        "stroke-linecap" -: "square"
        "stroke-miterlimit" -: "10"
        "shape-rendering" -: "geometricPrecision"

sidebarCSS :: Css
sidebarCSS = do
    "#fce" ? do
        height (px 40)
        width (pct 100)
        border solid (px 1) black
    "#fcecount" ? do
        width (pct 53)
        height (px 40)
        float floatLeft
        backgroundColor purple7
        display none
        textAlign $ alignSide sideCenter
        paddingLeft (px 15)
        paddingTop (px 5)
        fontSize (px 18)
    "#reset" ? do
        textAlign $ alignSide sideCenter
        float floatLeft
        width (pct 47)
        height (px 40)
        display none
        border solid (px 1) black
        cursor pointer
        ":hover" & do
            backgroundColor grey1
            fontColor white
    "#container" ? do
        width (pct 100)
        height (px 700)
        position relative
    "#sidebar" ? do
        display inlineBlock
        width (px 40)
        height (pct 100)
        float floatLeft
        backgroundColor purple8
        position absolute
        paddingLeft (px 23)
    "#sidebar-button" ? do
        cursor pointer
        display inlineBlock
        width (px 40)
        height100
        float floatLeft
        backgroundColor purple10
        position absolute
        border solid (px 1) black
        ":hover" & do
            backgroundColor purple6
    "#sidebar-icon" ? do
        width (px 30)
        height (px 50)
        paddingTop (px 20)
        paddingLeft (px 2)
        position absolute
        top (pct 40)
        left (px 4)
    "#focuses-nav, #graph-nav" ? do
        cursor pointer
    "#sidebar-nav" ? do
        width100
        fontSize (px 13)
        backgroundColor purple6
        border solid (px 1) grey2
        "box-shadow" -: "0 2px 2px -1px rgba(0, 0, 0, 0.055)"
        display block
        overflow hidden
        ul ? do
            width100
            margin0
            li ? do
                "list-style-type" -: "none"
                display inlineBlock
                width (pct 48)
                "-webkit-transition" -: "all 0.2s"
                "-moz-transition" -: "all 0.2s"
                "-ms-transition" -: "all 0.2s"
                "-o-transition" -: "all 0.2s"
                "transition" -: "all 0.2s"
                ":hover" & do
                    "background-color" -: "#5C497E !important"
                    a ? do
                        "color" -: "white !important"
                a ? do
                    color black
                    display inlineBlock
                    lineHeight (px 30)
                    alignCenter
                    width (pct 95)
                    textDecoration none
    "#focuses, #graphs" ? do
        marginTop (px 25)
        marginLeft (px 25)
        height100
        width100
        display none
    ".focus" ? do
        display block
        cursor pointer
        fontSize (px 20)
        border solid (px 1) black
        alignCenter
        width (pct 90)
        backgroundColor white
        "border-radius" -: "6px"
        ":hover" & do
            backgroundColor grey2
    "#close-focus" ? do
        display block
        cursor pointer
        backgroundColor purple6
        fontSize (px 20)
        border solid (px 1) black
        textAlign $ alignSide sideCenter
        width (pct 90)
    ".spotlight" & do
        semiVisible
        fill "white"
        stroke "none"
    ".details" & do
        border solid (px 1) black
        width (pct 90)
        height (px 0)
        marginBottom (px 7)
        overflow auto
        fontSize (px 14)
        fontColor white
        paddingLeft (px 5)
        paddingRight (px 5)
    ".active" & do
        backgroundColor purple8
    ".graph-button" & do
        display block
        cursor pointer
        fontSize (px 20)
        border solid (px 1) black
        "border-radius" -: "6px"
        textAlign $ alignSide sideCenter
        width (pct 90)
        backgroundColor white
        marginBottom (px 20)
        ":hover" & do
            backgroundColor grey2
    ".flip" & do
        transform $ scaleX (-1)


{- titleCSS
 - Generates CSS for the title. -}
titleCSS :: Css
titleCSS = "#svgTitle" ? do
    fontSize $ em 2.5
    fontWeight bold
    fontFamily ["Bitter"] [serif]
    fontStyle italic
    fill titleColour

{- regionCSS
 - Generates CSS for focus regions in the graph. -}
regionCSS :: Css
regionCSS = do
    "#region-labels > text" ? do
        fontSize (pt regionFontSize)
    ".region" ? do
        "fill-opacity" -: "0.25"
