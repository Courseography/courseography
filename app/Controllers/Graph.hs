module Controllers.Graph (graphResponse, index, graphImageResponse) where

import Control.Monad.IO.Class (liftIO)
-- import Data.Aeson (object, (.=))
-- import Data.Maybe (fromMaybe)
import Happstack.Server (Response, ServerPart, look, ok, toResponse)
import MasterTemplate (header, masterTemplate)
import Scripts (graphScripts)
import Text.Blaze ((!))
import qualified Text.Blaze.Html5 as H
import qualified Text.Blaze.Html5.Attributes as A

import Config (runDb)
-- import Database.CourseQueries (getGraph)
import Database.Persist.Sqlite (Entity, SelectOpt (Asc), SqlPersistM, selectList, (==.))
import Database.Tables as Tables (EntityField (GraphDynamic, GraphTitle), Graph, Text)
import Export.GetImages (getActiveGraphImage)
import Response.Image (returnImageData)
import Util.Happstack (createJSONResponse)

graphResponse :: ServerPart Response
graphResponse =
   ok $ toResponse $
    masterTemplate "Courseography - Graph"
                []
                (do
                    header "graph"
                    H.div ! A.id "container" $ ""
                )
                graphScripts


index :: ServerPart Response
index = liftIO $ runDb $ do
    graphsList :: [Entity Graph] <- selectList [GraphDynamic ==. False] [Asc GraphTitle]
    return $ createJSONResponse graphsList :: SqlPersistM Response


-- | Returns an image of the graph requested by the user, given graphInfo stored in local storage.
graphImageResponse :: ServerPart Response
graphImageResponse = do
    graphInfo <- look "JsonLocalStorageObj"
    (svgFilename, imageFilename) <- liftIO $ getActiveGraphImage graphInfo
    liftIO $ returnImageData svgFilename imageFilename
