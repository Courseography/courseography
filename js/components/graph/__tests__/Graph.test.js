import TestGraph from "./TestGraph"
import { fireEvent } from "@testing-library/react"
import { ZOOM_INCREMENT, KEYBOARD_PANNING_INCREMENT } from "../Graph"

describe("Graph Navigation", () => {
  it("Should pan right when the right arrow key is pressed", async () => {
    await TestGraph.build()
    let svg = document.querySelector("svg")
    let initialX = parseInt(svg.getAttribute("viewBox").split(" ")[0])
    fireEvent.keyDown(document.body, { key: "ArrowRight", code: "ArrowRight" })
    let newX = parseInt(svg.getAttribute("viewBox").split(" ")[0])
    let expected = initialX - KEYBOARD_PANNING_INCREMENT
    expect(newX).toBe(expected)
  })

  it("Should pan left when the left arrow key is pressed", async () => {
    await TestGraph.build()
    let svg = document.querySelector("svg")
    let initialX = parseInt(svg.getAttribute("viewBox").split(" ")[0])
    fireEvent.keyDown(document.body, { key: "ArrowLeft", code: "ArrowLeft" })
    let newX = parseInt(svg.getAttribute("viewBox").split(" ")[0])
    let expected = initialX + KEYBOARD_PANNING_INCREMENT
    expect(newX).toBe(expected)
  })

  it("Should pan down when the down arrow key is pressed", async () => {
    await TestGraph.build()
    let svg = document.querySelector("svg")
    let initialY = parseInt(svg.getAttribute("viewBox").split(" ")[1])
    fireEvent.keyDown(document.body, { key: "ArrowDown", code: "ArrowDown" })
    let newY = parseInt(svg.getAttribute("viewBox").split(" ")[1])
    let expected = initialY - KEYBOARD_PANNING_INCREMENT
    expect(newY).toBe(expected)
  })

  it("Should pan up when the up arrow key is pressed", async () => {
    await TestGraph.build()
    let svg = document.querySelector("svg")
    let initialY = parseInt(svg.getAttribute("viewBox").split(" ")[1])
    fireEvent.keyDown(document.body, { key: "ArrowUp", code: "ArrowUp" })
    let newY = parseInt(svg.getAttribute("viewBox").split(" ")[1])
    let expected = initialY + KEYBOARD_PANNING_INCREMENT
    expect(newY).toBe(expected)
  })

  it("Should zoom in when the user presses the + key", async () => {
    await TestGraph.build()
    let svg = document.querySelector("svg")
    let initialDims = svg
      .getAttribute("viewBox")
      .split(" ")
      .splice(2)
      .map(dim => parseFloat(dim))
    fireEvent.keyDown(document.body, { key: "+" })
    let newDims = svg
      .getAttribute("viewBox")
      .split(" ")
      .splice(2)
      .map(dim => parseFloat(dim))
    let expectedDims = initialDims.map(dim => dim * (1 - ZOOM_INCREMENT))

    expect(newDims[0]).toBe(expectedDims[0])
    expect(newDims[1]).toBe(expectedDims[1])
  })

  it("Should zoom out when the user presses the - key", async () => {
    await TestGraph.build()
    let svg = document.querySelector("svg")
    let initialDims = svg
      .getAttribute("viewBox")
      .split(" ")
      .splice(2)
      .map(dim => parseFloat(dim))
    fireEvent.keyDown(document.body, { key: "-" })
    let newDims = svg
      .getAttribute("viewBox")
      .split(" ")
      .splice(2)
      .map(dim => parseFloat(dim))
    let expectedDims = initialDims.map(dim => dim * (1 + ZOOM_INCREMENT))

    expect(newDims[0]).toBe(expectedDims[0])
    expect(newDims[1]).toBe(expectedDims[1])
  })

  it("Should pan when the user clicks and drags", async () => {
    await TestGraph.build()
    let svg = document.querySelector("svg")
    let initialX = parseInt(svg.getAttribute("viewBox").split(" ")[0])
    let initialY = parseInt(svg.getAttribute("viewBox").split(" ")[1])
    fireEvent.mouseDown(svg, { clientX: 250, clientY: 10 })
    fireEvent.mouseMove(svg, { clientX: 100, clientY: 20 })
    fireEvent.mouseUp(svg, { clientX: 100, clientY: 20 })
    let newX = parseInt(svg.getAttribute("viewBox").split(" ")[0])
    let newY = parseInt(svg.getAttribute("viewBox").split(" ")[1])
    let expectedX = initialX + 150
    let expectedY = initialY - 10

    expect(newX).toBe(expectedX)
    expect(newY).toBe(expectedY)
  })

  it("Should zoom in when the mouse wheel is scrolled down", async () => {
    const graph = await TestGraph.build()
    let svg = document.querySelector("svg")
    let initialDims = svg
      .getAttribute("viewBox")
      .split(" ")
      .splice(2)
      .map(dim => parseFloat(dim))
    let reactGraph = graph.getByTestId("react-graph")
    fireEvent.wheel(reactGraph, { deltaY: -1 })
    let newDims = svg
      .getAttribute("viewBox")
      .split(" ")
      .splice(2)
      .map(dim => parseFloat(dim))
    let expectedDims = initialDims.map(dim => dim * (1 - ZOOM_INCREMENT))
    expect(newDims[0]).toBe(expectedDims[0])
    expect(newDims[1]).toBe(expectedDims[1])
  })

  it("Should zoom out when the mouse wheel is scrolled up", async () => {
    const graph = await TestGraph.build()
    let svg = document.querySelector("svg")
    let initialDims = svg
      .getAttribute("viewBox")
      .split(" ")
      .splice(2)
      .map(dim => parseFloat(dim))
    let reactGraph = graph.getByTestId("react-graph")
    fireEvent.wheel(reactGraph, { deltaY: 1 })
    let newDims = svg
      .getAttribute("viewBox")
      .split(" ")
      .splice(2)
      .map(dim => parseFloat(dim))
    let expectedDims = initialDims.map(dim => dim * (1 + ZOOM_INCREMENT))
    expect(newDims[0]).toBe(expectedDims[0])
    expect(newDims[1]).toBe(expectedDims[1])
  })
})
