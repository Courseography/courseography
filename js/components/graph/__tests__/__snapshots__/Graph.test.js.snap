// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BoolGroup BoolGroup should match shallow snapshot 1`] = `
<g
  id="bools"
/>
`;

exports[`EdgeGroup should match shallow snapshot 1`] = `
<g
  id="edges"
/>
`;

exports[`Graph rendering RegionGroup should match shallow snapshot 1`] = `
<g
  id="regions"
>
  <path
    className="region"
    d="M17.386348,281.07376883 17.386348,14.41568908 316.385978,14.41576883 316.386348,281.07376883 "
    style={
      {
        "fill": "#6276b9",
      }
    }
  />
  <text
    className="region-label"
    style={
      {
        "fill": "#000000",
      }
    }
    x={1088.1677413999998}
    y={201.36453113000002}
  >
    Systems
  </text>
</g>
`;

exports[`NodeGroup should match shallow snapshot 1`] = `
<g
  id="nodes"
/>
`;
